### ISSUE TITLE
Invalid CS8609 when base method uses MaybeNull

### ISSUE THREAD

**Issue by YairHalberstadt** [dotnet/roslyn#41987]
**Version Used**: Visual Studio 16.5 preview 3

**Steps to Reproduce**:

```csharp
#nullable enable

using System.Diagnostics.CodeAnalysis;

public class Base<T>
{
	[return: MaybeNull]
	public virtual T M() => default;
}

public class Derived : Base<object>
{
	public override object? M() => default;
}
```

**Expected Behavior**: No Diagnostic

**Actual Behavior**: CS8609: Nullability of reference types in return type doesn't match overridden member.

**Comment by YairHalberstadt** [comment]
@jcouv is it possible you've fixed this as part of your recent work on these attributes?

**Comment by jcouv** [comment]
My recent changes would have mostly added warnings, not removed them.
This issue still reproes in 16.6p1.

Tagging @agocke who implemented the nullability variance.

**Comment by YairHalberstadt** [comment]
A workaround is to mark the override as `[return: MaybeNull]` and non-nullable.

**Comment by agocke** [comment]
I assumed this would be part of "making attributes part of method signatures" since this is part of OHI.

**Comment by YairHalberstadt** [comment]
@agocke

Are you saying that you would have expected this to work already? Is it a bug then?

What is OHI?

**Comment by agocke** [comment]
OHI = overriding, hiding, implementation

So, if a base method has `MaybeNull` on it, the override should also have `MaybeNull`. I know this work was approved, but I don't know its status. @jcouv?

**Comment by YairHalberstadt** [comment]
Currently the IDE generates MaybeNull when overriding a generic with a generic. It doesn't do that when overriding a generic with something non-generic. Nor should it, as it should use a nullable type instead.

**Comment by jcouv** [comment]
@agocke It is not necessary for the override to also have `MaybeNull`.
We do two checks, one without attributes (`object?` -> `T`, which seems like it should be okay, but I suspect produces the present warning) and one with attributes (not causing the problem here, as that would be a different warning).

**Comment by agocke** [comment]
@jcouv It should have `MaybeNull` if it's also generic, shouldn't it?

**Comment by jcouv** [comment]
@agocke That's a different scenario, but no, it's okay for the override to return a not-null value (ie. no `MaybeNull`). We just need to be able to assign from `overridingValue` to `overriddenType`.

**Comment by agocke** [comment]
Right, I forgot the example was return type. Use parameter type instead. The point being, there's a variance requirement in overloading that should be enforced.

**Comment by sharwell** [comment]
CS8609 should not be reported here for two reasons:

1. `MaybeNull` would be redundant
2. A derived type is allowed to return a non-nullable reference type while overriding a member that returns a nullable reference type (#23268)

Related to #41884

**Comment by YairHalberstadt** [comment]
@sharwell 

Here the derived type is a nullable reference type, so 2 doesn't apply.

**Comment by sharwell** [comment]
@YairHalberstadt Option 2 still applies in concept. It means that not only are the forms here equivalent, but the check itself is fundamentally incorrect because the language allows covariant types here.

**Comment by jcouv** [comment]
This was fixed by PR https://github.com/dotnet/roslyn/pull/42491
It contains very similar tests to this scenario, and I also verified this scenario manually. I'll go ahead and close. Thanks for reporting!
